File: backend/app/routers/organizations.py
from __future__ import annotations

import sqlite3
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from ..database import (
    _organization_slug,
    ensure_local_organization,
    get_db,
    record_audit_log,
)
from ..org_deps import get_membership
from . import auth as auth_module
from .auth import get_current_user


router = APIRouter(prefix="/organizations", tags=["organizations"])


def _require_authenticated(user: dict | None) -> None:
    if user is None and auth_module.AUTH_ENABLED:
        raise HTTPException(status_code=401, detail="请先登录")


class OrganizationCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    slug: str | None = Field(default=None, min_length=1, max_length=80)


def _organization_payload(row: sqlite3.Row, *, role: str | None = None) -> dict:
    result = {
        "id": row["id"],
        "name": row["name"],
        "slug": row["slug"],
        "status": row["status"],
        "owner_user_id": row["owner_user_id"],
        "settings": row["settings"],
        "created_at": row["created_at"],
    }
    if role is not None:
        result["role"] = role
    return result


@router.get("")
def list_organizations(
    connection: sqlite3.Connection = Depends(get_db),
    user: dict | None = Depends(get_current_user),
) -> list[dict]:
    _require_authenticated(user)
    if user is None:
        rows = connection.execute(
            "SELECT * FROM organizations WHERE status = 'active' ORDER BY created_at, id"
        ).fetchall()
        return [
            {**_organization_payload(row, role="owner"), "is_active": row["slug"] == "local"}
            for row in rows
        ]

    rows = connection.execute(
        """
        SELECT o.*, om.role, u.active_organization_id
        FROM organization_members AS om
        JOIN organizations AS o ON o.id = om.organization_id
        JOIN users AS u ON u.id = om.user_id
        WHERE om.user_id = ? AND om.status = 'active' AND o.status = 'active'
        ORDER BY o.created_at, o.id
        """,
        (int(user["id"]),),
    ).fetchall()
    return [
        {
            **_organization_payload(row, role=row["role"]),
            "is_active": int(row["id"]) == int(row["active_organization_id"] or 0),
        }
        for row in rows
    ]


@router.post("", status_code=201)
def create_organization(
    request: OrganizationCreate,
    connection: sqlite3.Connection = Depends(get_db),
    user: dict | None = Depends(get_current_user),
) -> dict:
    _require_authenticated(user)
    name = request.name.strip()
    if not name:
        raise HTTPException(status_code=422, detail="组织名称不能为空")
    owner_user_id = int(user["id"]) if user else None
    slug = _organization_slug(connection, request.slug or name)
    try:
        cursor = connection.execute(
            """
            INSERT INTO organizations(name, slug, owner_user_id, settings)
            VALUES (?, ?, ?, '{}')
            """,
            (name, slug, owner_user_id),
        )
    except sqlite3.IntegrityError as error:
        raise HTTPException(status_code=409, detail="组织标识已存在") from error
    organization_id = int(cursor.lastrowid)
    if owner_user_id is not None:
        connection.execute(
            """
            INSERT INTO organization_members(organization_id, user_id, role, status)
            VALUES (?, ?, 'owner', 'active')
            """,
            (organization_id, owner_user_id),
        )
        connection.execute(
            "UPDATE users SET active_organization_id = ? WHERE id = ?",
            (organization_id, owner_user_id),
        )
    record_audit_log(
        connection,
        organization_id=organization_id,
        actor_user_id=owner_user_id,
        action="organization.created",
        resource_type="organization",
        resource_id=organization_id,
    )
    connection.commit()
    row = connection.execute(
        "SELECT * FROM organizations WHERE id = ?", (organization_id,)
    ).fetchone()
    return _organization_payload(row, role="owner")


@router.get("/{organization_id}")
def get_organization(
    organization_id: int,
    connection: sqlite3.Connection = Depends(get_db),
    user: dict | None = Depends(get_current_user),
) -> dict:
    _require_authenticated(user)
    row = connection.execute(
        "SELECT * FROM organizations WHERE id = ? AND status = 'active'",
        (organization_id,),
    ).fetchone()
    if row is None:
        raise HTTPException(status_code=404, detail="组织不存在")
    role = "owner"
    if user is not None:
        membership = get_membership(connection, int(user["id"]), organization_id)
        if membership is None:
            raise HTTPException(status_code=403, detail="不是该组织成员")
        role = str(membership["role"])
    members = connection.execute(
        """
        SELECT u.id AS user_id, u.username, om.role, om.status, om.joined_at
        FROM organization_members AS om
        JOIN users AS u ON u.id = om.user_id
        WHERE om.organization_id = ? AND om.status = 'active'
        ORDER BY om.joined_at, u.id
        """,
        (organization_id,),
    ).fetchall()
    return {
        **_organization_payload(row, role=role),
        "members": [dict(member) for member in members],
    }


@router.post("/{organization_id}/switch")
def switch_organization(
    organization_id: int,
    connection: sqlite3.Connection = Depends(get_db),
    user: dict | None = Depends(get_current_user),
) -> dict:
    _require_authenticated(user)
    if user is None:
        row = connection.execute(
            "SELECT id FROM organizations WHERE id = ? AND status = 'active'",
            (organization_id,),
        ).fetchone()
        if row is None:
            raise HTTPException(status_code=404, detail="组织不存在")
        return {"switched": True, "organization_id": organization_id}
    membership = get_membership(connection, int(user["id"]), organization_id)
    if membership is None:
        raise HTTPException(status_code=403, detail="不是该组织成员")
    connection.execute(
        "UPDATE users SET active_organization_id = ? WHERE id = ?",
        (organization_id, int(user["id"])),
    )
    record_audit_log(
        connection,
        organization_id=organization_id,
        actor_user_id=int(user["id"]),
        action="organization.switched",
        resource_type="organization",
        resource_id=organization_id,
    )
    connection.commit()
    return {"switched": True, "organization_id": organization_id}

File: frontend/src/services/organizations.ts
import { reactive } from 'vue'
import { get, getOrganizationId, isOffline, post, setOrganizationId } from '../api'

export interface Organization {
  id: number
  name: string
  slug: string
  status: string
  owner_user_id: number | null
  settings: string
  created_at: string
  role?: string
  is_active?: boolean
  members?: Array<{
    user_id: number
    username: string
    role: string
    status: string
    joined_at: string
  }>
}

export const organizationsState = reactive({
  items: [] as Organization[],
  activeId: getOrganizationId(),
  loading: false,
  error: '',
})

const offlineOrganization: Organization = {
  id: 0,
  name: '本地学习空间',
  slug: 'local',
  status: 'active',
  owner_user_id: null,
  settings: '{}',
  created_at: '',
  role: 'owner',
  is_active: true,
}

export async function loadOrganizations(): Promise<Organization[]> {
  organizationsState.loading = true
  organizationsState.error = ''
  try {
    if (isOffline()) {
      organizationsState.items = [offlineOrganization]
      organizationsState.activeId = 0
      return organizationsState.items
    }
    organizationsState.items = await get<Organization[]>('/organizations')
    const active = organizationsState.items.find((item) => item.is_active)
      || organizationsState.items.find((item) => item.id === organizationsState.activeId)
      || organizationsState.items[0]
    organizationsState.activeId = active?.id ?? null
    setOrganizationId(organizationsState.activeId)
    return organizationsState.items
  } catch (error) {
    organizationsState.error = error instanceof Error ? error.message : String(error)
    throw error
  } finally {
    organizationsState.loading = false
  }
}

export async function switchOrganization(organizationId: number): Promise<void> {
  if (organizationId === organizationsState.activeId) return
  if (!isOffline()) await post(`/organizations/${organizationId}/switch`)
  organizationsState.activeId = organizationId
  setOrganizationId(organizationId)
  await loadOrganizations()
}

export async function createOrganization(name: string): Promise<Organization> {
  const organization = isOffline()
    ? offlineOrganization
    : await post<Organization>('/organizations', { name })
  if (isOffline()) {
    organizationsState.items = [organization]
    organizationsState.activeId = organization.id
  } else {
    await loadOrganizations()
  }
  return organization
}

